dimod/include/dimod/adjarraybqm.h

```diff
@@ -0,0 +1,156 @@
+// Copyright 2019 D-Wave Systems Inc.
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#ifndef DIMOD_ADJARRAYBQM_H_
+#define DIMOD_ADJARRAYBQM_H_
+
+#include <algorithm>
+#include <utility>
+#include <vector>
+
+namespace dimod {
+
+template<class V, class B>
+bool comp_v(const std::pair<V, B> ub, V v) {
+    return ub.first < v;
+}
+
+template<class V, class B, class N = std::size_t>
+class AdjArrayBQM {
+ public:
+    typedef B bias_type;
+    typedef N neighborhood_type;
+    typedef V variable_type;
+    typedef std::size_t size_type;
+
+    std::vector<std::pair<N, B>> invars;
+    std::vector<std::pair<V, B>> outvars;
```

Comment:
So I agree in general _except_ that some solvers will want direct access for various reasons. And since (I think) in c++ private means private, it makes sense to leave them public for now. In the future when we have more/better iterator access we can probably start hiding them